### PR TITLE
Create new app-operator rbac permissions

### PR DESCRIPTION
**Pull request by kubasobon**
Towards https://github.com/giantswarm/giantswarm/issues/21891

## Checklist

- [x] Update changelog in CHANGELOG.md.
- [x] If the change introduces new roles, please consider adding user-friendly descriptions to the roles with the annotation `giantswarm.io/notes` to help explain their purpose and usage.

### CODE REVIEW

File: service/controller/clusternamespace/resource/rbacappoperator/config.go [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+			{
+				APIGroups: []string{""},
+				Resources: []string{"configmaps"},
+				Verbs:     []string{"*"},
+			},
+			{
+				APIGroups: []string{""},
+				Resources: []string{"secrets"},
+				Verbs:     []string{"*"},
+			},
```
Comment (kubasobon) [inline service/controller/clusternamespace/resource/rbacappoperator/config.go:172]: The original role had some permissions for Events and app-operators still emit them in their own namespace.

```suggestion
			},
			{
				APIGroups: []string{""},
				Resources: []string{"events"},
				Verbs:     []string{"create", "patch", "update"},
			},
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Added
+
+- Remove `app-operator-*` ClusterRole and ClusterRoleBinding for old app-operator versions.
+
```
Comment (kubasobon) [inline CHANGELOG.md:13]: Please add changelog updates.

File: service/controller/clusternamespace/clusternamespace_resources.go [hunk 3]
```diff
@@ -32,8 +35,36 @@ func newClusterNamespaceResources(config clusterNamespaceResourcesConfig) ([]res
...
+		rbacAppOperatorResource, err = rbacappoperator.New(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
 	resources := []resource.Interface{
 		clusterNamespaceResourcesResource,
+		rbacCleanerResource,
+		rbacAppOperatorResource,
```
Comment (kubasobon) [inline service/controller/clusternamespace/clusternamespace_resources.go:67]: nit: sort

File: service/controller/clusternamespace/resource/clusternamespaceresources/create.go [hunk 4]
```diff
@@ -2,15 +2,15 @@ package clusternamespaceresources
...
 
 	"github.com/giantswarm/microerror"
 	security "github.com/giantswarm/organization-operator/api/v1alpha1"
 	rbacv1 "k8s.io/api/rbac/v1"
-	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/types"
 
+	"github.com/giantswarm/rbac-operator/pkg/rbac"
+
```
Comment (kubasobon) [inline service/controller/clusternamespace/resource/clusternamespaceresources/create.go:13]: Please remove the empty line and sort imports with `go fmt` or `goimports`.

File: service/controller/clusternamespace/resource/rbacappoperator/create.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+	r.logger.LogCtx(
+		ctx, "level", "info",
+		"message", fmt.Sprintf("Reconciling cluster namespace: %s.", cl.Name),
+	)
+
+	// Allow working with some generic resources across namespaces
+	err = r.CreateClusterRoleAndBinding(ctx, cl)
+	if err != nil {
+		return err
```
Comment (kubasobon) [inline service/controller/clusternamespace/resource/rbacappoperator/create.go:29]:
```suggestion
		return microerror.Mask(err)
```

File: service/controller/clusternamespace/resource/rbacappoperator/create.go [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+	}
+
+	// Allow getting catalog configmaps in giantswarm namespace
+	err = r.CreateCatalogReaderRoleAndBinding(ctx, cl)
+	if err != nil {
+		return err
```
Comment (kubasobon) [inline service/controller/clusternamespace/resource/rbacappoperator/create.go:35]:
```suggestion
		return microerror.Mask(err)
```

File: service/controller/clusternamespace/resource/rbacappoperator/create.go [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+	}
+
+	// Allow working with stuff in its own namespace
+	err = r.CreateOwnNamespaceRoleAndBinding(ctx, cl)
+	if err != nil {
+		return err
```
Comment (kubasobon) [inline service/controller/clusternamespace/resource/rbacappoperator/create.go:41]:
```suggestion
		return microerror.Mask(err)
```

File: service/controller/clusternamespace/resource/rbacappoperator/create.go [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (r *Resource) CreateCatalogReaderRoleAndBinding(ctx context.Context, ns corev1.Namespace) error {
+	var catalogReaderRole = getAppOperatorCatalogReaderRole(ns)
+
+	// TODO Move namespace to keys
```
Comment (kubasobon) [inline service/controller/clusternamespace/resource/rbacappoperator/create.go:66]: Is this `TODO` still important?

File: service/controller/clusternamespace/resource/rbacappoperator/create_test.go [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	clientgofake "k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/kubernetes/scheme"
+	clientfake "sigs.k8s.io/controller-runtime/pkg/client/fake"
+
+	"github.com/giantswarm/rbac-operator/service/controller/clusternamespace/key"
+)
+
+func Test_EnsureCreated(t *testing.T) {
```
Comment (kubasobon) [inline service/controller/clusternamespace/resource/rbacappoperator/create_test.go:22]: :+1: :guitar:

File: service/controller/rbac/resource/externalresources/create.go [hunk 10]
```diff
@@ -2,11 +2,11 @@ package externalresources
 
 import (
 	"context"
-	"fmt"
+
+	"github.com/giantswarm/rbac-operator/pkg/rbac"
```
Comment (kubasobon) [inline service/controller/rbac/resource/externalresources/create.go:6]: Please group this import with other `rbac-operator` imports.

File: service/internal/bootstrap/resources.go [hunk 11]
```diff
@@ -7,6 +7,8 @@ import (
 	"context"
 	"fmt"
 
+	"github.com/giantswarm/rbac-operator/pkg/rbac"
```
Comment (kubasobon) [inline service/internal/bootstrap/resources.go:10]: ditto

**Review by kubasobon — COMMENTED** [review]
Can't approve because I'm an author as well :sweat_smile: 

Looks good, left some comments. Let's test it again before merging.

**Comment by kubasobon** [comment]
@uvegla
>  If the change introduces new roles, please consider adding user-friendly descriptions to the roles with the annotation `giantswarm.io/notes` to help explain their purpose and usage.

It would be a great addition to all the RBAC resources we set up.

File: service/controller/clusternamespace/resource/rbaccleaner/resource.go [hunk 12]
```diff
@@ -0,0 +1,51 @@
+// TODO: rbaccleaner
```
Comment (MarcelMue) [inline service/controller/clusternamespace/resource/rbaccleaner/resource.go]: Missing removal?

File: service/controller/clusternamespace/resource/rbacappoperator/delete_test.go [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+			// Execute
+
+			err = resource.EnsureDeleted(context.TODO(), wcNamespace)
+			if err != nil {
+				t.Fatalf("error == %#v, want nil", err)
+			}
+
+			// Checks
+			_, err = k8sClientFake.K8sClient().RbacV1().ClusterRoles().Get(context.TODO(), clusterRole.Name, metav1.GetOptions{})
+			if !errors.IsNotFound(err) {
```
Comment (MarcelMue) [inline service/controller/clusternamespace/resource/rbacappoperator/delete_test.go:99]: Small input: Would it make sense to also print the error here? E.g. in `t.Fatalf` we should include the error if it's not NotFound.

File: service/controller/clusternamespace/clusternamespace_resources.go (hunk 3 above)
Comment (uvegla) [reply service/controller/clusternamespace/clusternamespace_resources.go:66]: @kubasobon What do you mean? Is the updated formatting / order better / what you intended?

File: service/controller/clusternamespace/resource/rbaccleaner/create.go [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+
+	"github.com/giantswarm/rbac-operator/pkg/rbac"
+
+	"github.com/giantswarm/microerror"
+
+	"github.com/giantswarm/rbac-operator/service/controller/clusternamespace/key"
+)
+
+// EnsureCreated Ensures that ClusterRoleBinding and ClusterRole deployed by legacy
+// app-operators to Cluster namespaces are *deleted* when a cluster namespace is created.
```
Comment (ljakimczuk) [inline service/controller/clusternamespace/resource/rbaccleaner/create.go:14]: Do we have something similar for the org- namespaces?

File: service/controller/clusternamespace/resource/rbaccleaner/create.go [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+// We will be using other roles and cluster roles created by another resource.
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	var err error
+
+	cl, err := key.ToNamespace(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	nameToDelete := key.AppOperatorClusterRoleNameFromNamespace(cl)
```
Comment (ljakimczuk) [inline service/controller/clusternamespace/resource/rbaccleaner/create.go:24]: What about the `%s-app-operator` ClusterRole(Binding) from org- namespaces?

**Review by ljakimczuk — APPROVED** [review]

File: service/controller/clusternamespace/resource/rbaccleaner/create.go (hunk 14 above)
Comment (kubasobon) [reply service/controller/clusternamespace/resource/rbaccleaner/create.go:14]: I checked and we don't have app-operators in the org- namespace in any MC. I think the move of resources to org- namespaces goes hand in hand with CAPX, so it's cluster-apps-operator running in there.

Comment (kubasobon) [reply service/controller/clusternamespace/resource/rbaccleaner/create.go:14]: @MarcelMue might know if I'm totally wrong or not.

File: service/controller/clusternamespace/resource/rbacappoperator/delete_test.go (hunk 13 above)
Comment (uvegla) [reply service/controller/clusternamespace/resource/rbacappoperator/delete_test.go:99]: Makes sense, updated! 👍🏻

**Review by anvddriesch — APPROVED** [review]
LGTM if tested but please make sure to test including cluster deletion.

**Comment by uvegla** [comment]
> LGTM if tested but please make sure to test including cluster deletion.

Tested, the `rbac-operator` managed rbac resources are cleaned up when the cluster namespace gets deleted.

```
I 05/12 12:29:49 /v9om6 rbacappoperator Cleaning up app-operator resources for cluster namespace: v9om6. | rbac-operator/service/controller/clusternamespace/resource/rbacappoperator/delete.go:19 | controller=rbac-operator-cluster-namespace-controller | event=delete | loop=240 | version=613218695
I 05/12 12:29:49 /v9om6 rbacappoperator Deleting ClusterRole app-operator-v9om6-by-rbac-operator | rbac-operator/pkg/rbac/cluster_role.go:55 | controller=rbac-operator-cluster-namespace-controller | event=delete | loop=240 | version=613218695
I 05/12 12:29:49 /v9om6 rbacappoperator ClusterRole app-operator-v9om6-by-rbac-operator has been deleted. | rbac-operator/pkg/rbac/cluster_role.go:63 | controller=rbac-operator-cluster-namespace-controller | event=delete | loop=240 | version=613218695
I 05/12 12:29:49 /v9om6 rbacappoperator Deleting ClusterRoleBinding app-operator-v9om6-by-rbac-operator | rbac-operator/pkg/rbac/cluster_role_binding.go:70 | controller=rbac-operator-cluster-namespace-controller | event=delete | loop=240 | version=613218695
I 05/12 12:29:49 /v9om6 rbacappoperator ClusterRoleBinding app-operator-v9om6-by-rbac-operator has been deleted. | rbac-operator/pkg/rbac/cluster_role_binding.go:78 | controller=rbac-operator-cluster-namespace-controller | event=delete | loop=240 | version=613218695
I 05/12 12:29:49 /v9om6 rbacappoperator Deleting Role `app-operator-v9om6-by-rbac-operator` in namespace giantswarm. | rbac-operator/pkg/rbac/role.go:70 | controller=rbac-operator-cluster-namespace-controller | event=delete | loop=240 | version=613218695
I 05/12 12:29:49 /v9om6 rbacappoperator Role `app-operator-v9om6-by-rbac-operator` in namespace giantswarm has been deleted. | rbac-operator/pkg/rbac/role.go:78 | controller=rbac-operator-cluster-namespace-controller | event=delete | loop=240 | version=613218695
I 05/12 12:29:49 /v9om6 rbacappoperator Deleting RoleBinding `app-operator-v9om6-by-rbac-operator` in namespace giantswarm. | rbac-operator/pkg/rbac/role_binding.go:69 | controller=rbac-operator-cluster-namespace-controller | event=delete | loop=240 | version=613218695
I 05/12 12:29:49 /v9om6 rbacappoperator RoleBinding `app-operator-v9om6-by-rbac-operator` in namespace giantswarm has been deleted. | rbac-operator/pkg/rbac/role_binding.go:77 | controller=rbac-operator-cluster-namespace-controller | event=delete | loop=240 | version=613218695
I 05/12 12:29:49 /v9om6 rbacappoperator Deleting Role `app-operator-v9om6-by-rbac-operator` in namespace v9om6. | rbac-operator/pkg/rbac/role.go:70 | controller=rbac-operator-cluster-namespace-controller | event=delete | loop=240 | version=613218695
I 05/12 12:29:49 /v9om6 rbacappoperator Role `app-operator-v9om6-by-rbac-operator` in namespace v9om6 has been deleted. | rbac-operator/pkg/rbac/role.go:78 | controller=rbac-operator-cluster-namespace-controller | event=delete | loop=240 | version=613218695
I 05/12 12:29:49 /v9om6 rbacappoperator Deleting RoleBinding `app-operator-v9om6-by-rbac-operator` in namespace v9om6. | rbac-operator/pkg/rbac/role_binding.go:69 | controller=rbac-operator-cluster-namespace-controller | event=delete | loop=240 | version=613218695
I 05/12 12:29:49 /v9om6 rbacappoperator RoleBinding `app-operator-v9om6-by-rbac-operator` in namespace v9om6 has been deleted. | rbac-operator/pkg/rbac/role_binding.go:77 | controller=rbac-operator-cluster-namespace-controller | event=delete | loop=240 | version=613218695
```

File: service/controller/clusternamespace/clusternamespace_resources.go (hunk 3 above)
Comment (kubasobon) [reply service/controller/clusternamespace/clusternamespace_resources.go:66]: Ah, the resources were out of alphabetical order, but it's a microscopically tiny nit. Feel free to ignore :D

**Review by MarcelMue — APPROVED** [review]

File: CHANGELOG.md [hunk 16]
```diff
@@ -7,9 +7,17 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Added
+
+- Add new roles and cluster roles to app-operators when a cluster namespace is created and clean them up when the cluster namespace is destroyed
+
+## Removed
```
Comment (marians) [inline CHANGELOG.md:14]:
```suggestion
### Removed
```

File: pkg/label/label.go [hunk 17]
```diff
@@ -7,8 +7,10 @@ const (
 
 	ManagedBy = "giantswarm.io/managed-by"
 
-	// Labels, used in legacy cluster namespaces
+	// LegacyCustomer Labels, used in legacy cluster namespaces
```
Comment (marians) [inline pkg/label/label.go:10]:
```suggestion
	// LegacyCustomer labels, used in legacy cluster namespaces
```

File: pkg/label/label.go [hunk 18]
```diff
@@ -7,8 +7,10 @@ const (
...
 	LegacyCustomer = "customer"
+
+	NOTES = "giantswarm.io/notes"
```
Comment (marians) [inline pkg/label/label.go:13]: Coud we use https://github.com/giantswarm/k8smetadata/blob/dff3b2358977e13c90479c66e21c728a96ddfe6d/pkg/annotation/general.go#L12 here?

Comment (uvegla) [reply pkg/label/label.go:13]: Cool, did not know about them, thanks! :)

**Pull request merged**